nameresolution/consul/watcher.go

```diff
@@ -0,0 +1,167 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	consul "github.com/hashicorp/consul/api"
+)
+
+const (
+	// retryInterval is the base retry value
+	retryInterval = 5 * time.Second
+
+	// maximum back off time, this is to prevent exponential runaway
+	maxBackoffTime = 180 * time.Second
+)
+
+type watchPlan struct {
+	expired      bool
+	lastParamVal blockingParamVal
+	lastResult   []*consul.ServiceEntry
+	service      string
+	options      *consul.QueryOptions
+}
+
+type blockingParamVal interface {
+	equal(other blockingParamVal) bool
+	next(previous blockingParamVal) blockingParamVal
+}
+
+type waitIndexVal uint64
+
+// Equal implements BlockingParamVal
+func (idx waitIndexVal) equal(other blockingParamVal) bool {
+	if otherIdx, ok := other.(waitIndexVal); ok {
+		return idx == otherIdx
+	}
+
+	return false
+}
+
+// Next implements BlockingParamVal
+func (idx waitIndexVal) next(previous blockingParamVal) blockingParamVal {
+	if previous == nil {
+		return idx
+	}
+	prevIdx, ok := previous.(waitIndexVal)
+	if ok && prevIdx == idx {
+		// This value is the same as the previous index, reset
+		return waitIndexVal(0)
+	}
+
+	return idx
+}
+
+func (r *resolver) watch(p *watchPlan) (blockingParamVal, []*consul.ServiceEntry, bool, error) {
```

Comment:
Is this to keep the cache warm?

Reply:
correct, this is using blocking queries to keep relevant services warm in cache - see https://www.consul.io/api-docs/features/blocking